packages/documentation/copy/en/modules-reference/Theory.md

```diff
@@ -0,0 +1,415 @@
+
+```mermaid
+graph LR
+  main.ts -- "#quot;./math#quot;" --> math.ts
+```
+
+This isn’t entirely wrong, but the reality is deeper. The resolution of `"./math"` (and subsequently, the type of `add`) need to reflect the reality of what happens at runtime to the _output_ files. A more robust way to think about this process would look like this:
+
+```mermaid
+graph LR
+  subgraph Output files
+    main.js
+    style main.js stroke-dasharray: 5 5
+    math.js
+    style math.js stroke-dasharray: 5 5
+  end
+  subgraph Input files
+    main.ts
+    math.ts
+  end
+  main.ts -. Map to output .-> main.js
+  main.js -- "#quot;./math#quot;" --> math.js
+  math.js -. Map to input .-> math.ts
+```
+
+This model makes it clear that for TypeScript, module resolution is mostly a matter of accurately modeling the host’s module resolution algorithm between output files, with a little bit of remapping applied to find type information. Let’s look at another example that appears unintuitive through the lens of the simple model, but makes perfect sense with the robust model:
+
+```ts
+// @moduleResolution: node16
+// @rootDir: src
+// @outDir: dist
+
+// @Filename: src/math.mts
+export function add(a: number, b: number) {
+  return a + b;
+}
+
+// @Filename: src/main.mts
+import { add } from "./math.mjs";
+add(1, 2);
+```
+
+Node.js ESM `import` declarations use a strict module resolution algorithm that requires relative paths to include file extensions. When we only think about input files, it’s a little strange that `"./math.mjs"` seems to resolve to `math.mts`. Since we’re using an `outDir` to put compiled outputs in a different directory, `math.mjs` doesn’t even exist next to `main.mts`! Why should this resolve? With our new mental model, it’s no problem:
+
+```mermaid
+graph LR
+  subgraph Output files
+    dist/main.mjs
+    style dist/main.mjs stroke-dasharray: 5 5
+    dist/math.mjs
+    style dist/math.mjs stroke-dasharray: 5 5
+  end
+  subgraph Input files
+    src/main.mts
+    src/math.mts
+  end
+  src/main.mts -. Map to output .-> dist/main.mjs
+  dist/main.mjs -- "#quot;./math.mjs#quot;" --> dist/math.mjs
+  dist/math.mjs -. Map to input .-> src/math.mts
+```
+
+Understanding this mental model may not immediately eliminate the strangeness of seeing output file extensions in input files, and it’s natural to think in terms of shortcuts: “`"./math.mjs"` refers to the input file `math.mts`. I have to write the output extension, but the compiler knows to look for `.mts` when I write `.mjs`.” This shortcut is even how the compiler works internally, but the more robust mental model explains _why_ module resolution in TypeScript works this way: given the constraint that the module specifier in the output file will be [the same](#module-specifiers-are-not-transformed) as the module specifier in the input file, this is the only process that accomplishes our two goals of validating output files and assigning types.
```

Comment:
This sentence felt a bit iffy because of the intermixed double quotes, it could be solved by italicising?

```suggestion
Understanding this mental model may not immediately eliminate the strangeness of seeing output file extensions in input files, and it’s natural to think in terms of shortcuts: _“`"./math.mjs"` refers to the input file `math.mts`. I have to write the output extension, but the compiler knows to look for `.mts` when I write `.mjs`.”_ This shortcut is even how the compiler works internally, but the more robust mental model explains _why_ module resolution in TypeScript works this way: given the constraint that the module specifier in the output file will be [the same](#module-specifiers-are-not-transformed) as the module specifier in the input file, this is the only process that accomplishes our two goals of validating output files and assigning types.
```